chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,300 @@
+    }
+}
+
+impl<E: 'static + Send + Sync> Service<E> {
+    /// Checks if the referred scope has been terminated.
+    pub fn is_terminated(&self) -> bool {
+        self.0.upgrade().is_none()
+    }
+
+    /// Waits until the scope is terminated.
+    /// Returns `ErrCanceled` iff `ctx` was canceled before that.
+    pub fn terminated<'a>(
+        &'a self,
+        ctx: &'a Ctx,
+    ) -> impl Future<Output = OrCanceled<()>> + Send + Sync + 'a {
+        let terminated = self.0.upgrade().map(|m| m.lock().unwrap().terminated.clone());
+        async move {
+            if let Some(t) = terminated {
+                ctx.wait(t.recv()).await
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Cancels the scope's context and waits until the scope is terminated.
+    /// Note that ErrCanceled is returned if the `ctx` passed as argument is canceled before that,
+    /// not when scope's context is cancelled.
+    pub fn terminate<'a>(
+        &'a self,
+        ctx: &'a Ctx,
+    ) -> impl Future<Output = OrCanceled<()>> + Send + Sync + 'a {
+        let terminated = self.0.upgrade().map(|m| {
+            let inner = m.lock().unwrap();
+            inner.ctx.cancel();
+            inner.terminated.clone()
+        });
+        async move {
+            if let Some(t) = terminated {
+                ctx.wait(t.recv()).await
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Spawns a task on this scope. Returns ErrTerminated if the scope has already terminated.
+    /// TODO(gprusak): consider returning a handle to the task, which can be then explicitly
+    /// awaited.
+    pub fn spawn(&self, f: impl AsyncFn<'static, Ctx, Result<(), E>>) -> Result<(), ErrTerminated> {
+        self.0.upgrade().map(|m| Inner::spawn(m, f.wrap())).ok_or(ErrTerminated)
+    }
+
+    pub fn new_service(&self) -> Result<Service<E>, ErrTerminated> {
+        self.0.upgrade().map(|m| Inner::new_service(m)).ok_or(ErrTerminated)
+    }
+}
+
+/// Wrapper of a scope reference which cancels the scope when dropped.
+/// Used by Scope to cancel the scope as soon as all tasks spawned via
+/// `Scope::spawn` complete.
+struct StrongService<E: 'static>(Arc<Mutex<Inner<E>>>);
+
+impl<E: 'static> Borrow<Mutex<Inner<E>>> for StrongService<E> {
+    fn borrow(&self) -> &Mutex<Inner<E>> {
+        &*self.0
+    }
+}
+
+impl<E: 'static> Drop for StrongService<E> {
+    fn drop(&mut self) {
+        self.0.lock().unwrap().ctx.cancel()
+    }
+}
+
+/// Scope represents a concurrent computation bounded by lifetime 'env.
+/// It should be created only via `run!` macro.
+/// Scope is cancelled when the provided context is cancelled.
+/// Scope is cancelled when any of the tasks in the scope returns an error.
+/// Scope is cancelled when all the tasks in the scope complete.
+/// Scope is terminated when it is cancelled AND all tasks in the scope complete.
+pub struct Scope<'env, E: 'static>(
```

Comment:
The `'env` lifetime here and the way it is utilized is confusing me. On one hand, users can create a `Scope<'_>` for arbitrary floating lifetime via `new_scope<'_>`. On another this float then gets restricted somewhat in practical use by an occasional use of `&'env Scope<'env>`. That interacts in some way with the lifetime of spawned `AsyncFn`s too in ways I didn’t yet wrap my head around. Admittedly I had a pretty poor night’s sleep T_T.

Do I understand correctly this is intended to similar to crossbeam’s [scope](https://docs.rs/crossbeam/latest/crossbeam/fn.scope.html)?… This code seems like it is attempting to replicate something similar, but then appears to be doing something slightly, but meaningfully different.

Anyhow, it would really help if you had a description of what each of the lifetimes used is intended to represent. Even better if you had some examples of what you expect to _not_ compile. It would be easier to see if the approach taken actually achieves the intended purpose with these in hand.

---

P.S. Have you seen https://docs.rs/async-scoped/latest/async_scoped/index.html? It does not support deadlines, but perhaps building some of that functionality on top of that would be easier? And/or it might have ideas you can borrow :)

Reply:
I've added some explanation at the top of the scope.rs . I'll work on positive and negative examples. async_scoped doesn't address my use case because although it awaits async tasks, it blocks the thread on which it is run. In contrast I'm implementing an async scope in an async environment.

Reply:
Okay, I’ll give another shot at understanding this this Friday (tomorrow is day off)

Reply:
So, I spent some time thinking about it, and I think this `'env` business is roughly equivalent to for the purposes of this functionality:

```diff
diff --git a/chain/network/src/concurrency/scope/mod.rs b/chain/network/src/concurrency/scope/mod.rs
index 8ebd760d8..1f32fec9c 100644
--- a/chain/network/src/concurrency/scope/mod.rs
+++ b/chain/network/src/concurrency/scope/mod.rs
@@ -295,18 +295,18 @@ impl<E: 'static> Drop for StrongService<E> {
 /// Scope is cancelled when any of the tasks in the scope returns an error.
 /// Scope is cancelled when all the tasks in the scope complete.
 /// Scope is terminated when it is cancelled AND all tasks in the scope complete.
-pub struct Scope<'env, E: 'static>(
+pub struct Scope<E: 'static>(
     /// Scope is equivalent to a strong service, but bounds
     Weak<StrongService<E>>,
-    /// Makes Scope<'env,E> invariant in 'env.
-    std::marker::PhantomData<fn(&'env ()) -> &'env ()>,
+    // Makes Scope<'env,E> invariant in 'env.
+    // std::marker::PhantomData<fn(&'env ()) -> &'env ()>,
 );
 
-impl<'env, E: 'static + Send + Sync> Scope<'env, E> {
-    pub fn spawn(&self, f: impl AsyncFn<'env, Ctx, Result<(), E>>) {
+impl<E: 'static + Send + Sync> Scope<E> {
+    pub fn spawn<'scope>(&'scope self, f: impl AsyncFn<'scope, Ctx, Result<(), E>>) {
         let f = f.wrap();
         let f =
-            unsafe { std::mem::transmute::<BoxAsyncFn<'env, _, _>, BoxAsyncFn<'static, _, _>>(f) };
+            unsafe { std::mem::transmute::<BoxAsyncFn<'scope, _, _>, BoxAsyncFn<'static, _, _>>(f) };
         Inner::spawn(self.0.upgrade().unwrap(), f);
     }
 
@@ -347,19 +347,19 @@ impl Drop for MustCompleteGuard {
 pub mod internal {
     use super::*;
 
-    pub fn new_scope<'env, E: 'static>() -> Scope<'env, E> {
-        Scope(Weak::new(), std::marker::PhantomData)
+    pub fn new_scope<E: 'static>() -> Scope<E> {
+        Scope(Weak::new())
     }
 
-    pub async fn run<'env, E, F, G>(
-        scope: &'env mut Scope<'env, E>,
+    pub async fn run<'scope, E, F, G>(
+        scope: &'scope mut Scope<E>,
         ctx: &Ctx,
         f: F,
     ) -> Result<(), E>
     where
         E: 'static + Send + Sync,
-        F: 'env + FnOnce(&'env Scope<'env, E>) -> G,
-        G: AsyncFn<'env, Ctx, Result<(), E>>,
+        F: FnOnce(&'scope Scope<E>) -> G,
+        G: AsyncFn<'scope, Ctx, Result<(), E>>,
     {
         must_complete(async move {
             let (inner, recv) = Inner::new(ctx);

```

With the most important change being to the `spawn` function. Which to my eyes is much simpler, but maybe I’m missing a detail that you thought about?

Reply:
A reference to Scope is covariant in its lifetime, so your example would allow spawning tasks with a lifetime shorter than the lifetime of the Scope. This would be unsound, because if you spawn from task A, a task B which references local variables of task A, then there is no guarantee that those local variables will life long enough (both A and B are guaranteed to terminate within lifetime of the Scope, but that's all.